primitives/relayers/src/lane_relayers.rs

```diff
@@ -0,0 +1,625 @@
+/// has not delivered any messages during previous epoch.
+///
+/// Relayers are bargaining for the place in the set by offering lower reward for delivering
+/// messages. Relayer, which agrees to get a lower reward will likely to replace a "more greedy"
+/// relayer in the `next_set`.
+#[derive(CloneNoBound, Decode, Encode, Eq, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(MaxNextRelayersPerLane))]
+pub struct NextLaneRelayersSet<
+	AccountId: Clone,
+	BlockNumber: Clone,
+	MaxNextRelayersPerLane: Get<u32>,
+> {
+	/// Number of block, where the active set may be replaced with the [`Self::next_set`].
+	///
+	/// We do not allow immediate changes of the active set, because relayers
+	/// may change it so that they are always assigned at the current slot.
+	may_enact_at: BlockNumber,
+	/// Next set of lane relayers.
+	///
+	/// It is a bounded priority queue. Relayers that are working for larger reward are replaced
+	/// with relayers, that are working for smaller reward.
+	next_set: BoundedVec<RelayerAndReward<AccountId>, MaxNextRelayersPerLane>,
+}
+
+impl<AccountId, BlockNumber, MaxNextRelayersPerLane>
+	NextLaneRelayersSet<AccountId, BlockNumber, MaxNextRelayersPerLane>
+where
+	AccountId: Clone + PartialOrd,
+	BlockNumber: Copy,
+	MaxNextRelayersPerLane: Get<u32>,
+{
+	/// Creates new empty relayers set, where next sets enacts at given block.
+	pub fn empty(may_enact_at: BlockNumber) -> Self {
+		NextLaneRelayersSet { may_enact_at, next_set: BoundedVec::new() }
+	}
+
+	/// Returns block, starting from which the `next_set` may be enacted.
+	pub fn may_enact_at(&self) -> BlockNumber {
+		self.may_enact_at
+	}
+
+	/// Set block, starting from which the `next_set` may be enacted.
+	pub fn set_may_enact_at(&mut self, may_enact_at: BlockNumber) {
+		self.may_enact_at = may_enact_at;
+	}
+
+	/// Returns relayer entry from the next set.
+	pub fn relayer(&self, relayer: &AccountId) -> Option<&RelayerAndReward<AccountId>> {
+		self.next_set.iter().find(|r| r.relayer() == relayer)
+	}
+
+	/// Returns relayers from the next set.
+	pub fn relayers(&self) -> &[RelayerAndReward<AccountId>] {
+		self.next_set.as_slice()
+	}
+
+	/// Try insert relayer to the next set.
+	///
+	/// Returns `true` if relayer has been added to the set and false otherwise.
+	pub fn try_push(
+		&mut self,
+		relayer: AccountId,
+		relayer_reward_per_message: RelayerRewardAtSource,
+	) -> bool {
+		// first, remove existing entry for the same relayer from the set
+		self.try_remove(&relayer);
+		// now try to insert new entry into the queue
+		self.next_set
+			.force_insert_keep_left(
+				self.select_position_in_next_set(relayer_reward_per_message),
+				RelayerAndReward { relayer, relayer_reward_per_message },
+			)
+			.is_ok()
+	}
+
+	/// Try remove relayer from the next set.
+	///
+	/// Returns `true` if relayer has been removed from the set.
+	pub fn try_remove(&mut self, relayer: &AccountId) -> bool {
```

Comment:
For consistency with `HashMap`, could we call this method `remove()` and return an Option<&RelayerAndReward> ?

Reply:
Renamed to `try_remove`. I guess you mean `Option<RelayerAndReward>`, not the `Option<&RelayerAndReward>`. Also done